### PR TITLE
Solve I_S queries using CNF rewriting

### PR DESCRIPTION
**Pull request by systay**
## Description

To handle I_S queries correctly, we need to be able to find the `table_schema` and `table_name` predicates in the WHERE clause. Sometimes, this is difficult because of how the query is written.

What this PR does is to introduce a rewriter that is able to massage boolean logic into a simpler form. It normalises the predicates into a form that the planner is looking for, and this makes it possible for the planner to find more applicable rules.

CNF, short for conjuctive normal form, rewrites the predicates into `an AND of ORs`, which is exactly what the planner is looking for.

## Related Issue(s)
Fixes: #7377

## Checklist
- [x] Should this PR be backported? `Nope`
- [x] Tests were added or are not required `Added`
- [x] Documentation was added or is not required `Not needed`

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving

### CODE REVIEW

**Comment by systay** [comment]
@pontusmelke am I doing it right?

**Comment by systay** [comment]
We could use this strategy in more situations - we could replan when we encounter Scatter queries, vtgate joins and probably some more.

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/planbuilder/plan_test.go [hunk 1]
```diff
@@ -611,51 +610,45 @@ func BenchmarkPlanner(b *testing.B) {
 		b.Run(filename+"-v3", func(b *testing.B) {
 			benchmarkPlanner(b, V3, testCases, vschema)
 		})
-		b.Run(filename+"-v4", func(b *testing.B) {
-			benchmarkPlanner(b, Gen4, testCases, vschema)
-		})
-		b.Run(filename+"-v4left2right", func(b *testing.B) {
-			benchmarkPlanner(b, Gen4Left2Right, testCases, vschema)
-		})
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/plan_test.go:619]: nope

File: go/vt/vtgate/planbuilder/plan_test.go [hunk 2]
```diff
@@ -611,51 +610,45 @@ func BenchmarkPlanner(b *testing.B) {
...
-	b.Run("Select (random sample, N=32)", func(b *testing.B) {
-		benchmarkPlanner(b, V3, selectCases[:32], vschema)
-	})
-}
+//func BenchmarkSelectVsDML(b *testing.B) {
+//	vschema := &vschemaWrapper{
+//		v:             loadSchema(b, "schema_test.json"),
+//		sysVarEnabled: true,
+//		version:       V3,
+//	}
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/plan_test.go:621]: oops

**Comment by pontusmelke** [comment]
> @pontusmelke am I doing it right? ;)

Still got it 💥

**Pull request merged**